### ISSUE TITLE
Getting 'does not specify a 'report-uri'' when actually do

### ISSUE THREAD

**Issue by Xosmond** [twitter/secureheaders#276]
# Getting 'does not specify a 'report-uri'' when actually do

Getting error message when receiving pdf document.
rails 5.0.0
secure_headers 3.3.2

### Expected outcome

```
No error message
```

### Actual outcome

```
The Content Security Policy 'plugin-types application/pdf' was delivered in report-only mode, but does not specify a 'report-uri'; the policy will have no effect. Please either add a 'report-uri' directive, or deliver the policy via the 'Content-Security-Policy' header.
```

### Config

```ruby
SecureHeaders::Configuration.default do |config|
  config.cookies = {
    secure: true, # mark all cookies as "Secure"
    httponly: true, # mark all cookies as "HttpOnly"
    samesite: {
      strict: true # mark all cookies as SameSite=Strict
    }
  }
  config.hsts = "max-age=#{20.years.to_i}; includeSubdomains; preload"
  config.x_frame_options = "SAMEORIGIN"
  config.x_content_type_options = "nosniff"
  config.x_xss_protection = "1; mode=block"
  config.x_download_options = "noopen"
  config.x_permitted_cross_domain_policies = "none"
  config.referrer_policy = "origin-when-cross-origin"
  config.csp = {
    # "meta" values. these will shaped the header, but the values are not included in the header.
    report_only: true,      # default: false
    preserve_schemes: true, # default: false. Schemes are removed from host sources to save bytes and discourage mixed content.

    # directive values: these values will directly translate into source directives
    default_src: %w(https: 'self'),
    frame_src: %w('self'),
    connect_src: %w('self' wws:),
    font_src: %w('self' data:),
    img_src: %w('self' www.gravatar.com data:),
    media_src: %w('self'),
    object_src: %w('self'),
    script_src: %w('self'),
    style_src: %w('self'),
    base_uri: %w('self'),
    child_src: %w('self'),
    form_action: %w('self' github.com),
    frame_ancestors: %w('self'),
    plugin_types: %w(application/pdf),
    report_uri: %w(https://[FILTERED].report-uri.io/r/default/csp/enforce)
  }
  config.hpkp = {
    report_only: false,
    max_age: 60.days.to_i,
    include_subdomains: true,
    plugin_types: %w(application/pdf),
    report_uri: "https://[FILTERED].report-uri.io/r/default/hpkp/enforce",
    pins: [
      {sha256: "abc"},
      {sha256: "123"}
    ]
  }
end
```

I don't call 'SecureHeaders::Configuration.override'

### Generated headers

HTTP/1.1 200 OK
Location: http://localhost:3000/sells
Content-Type: text/html; charset=utf-8
Cache-Control: no-cache
Set-Cookie: [FILTERED]_session=[FILTERED]; path=/; HttpOnly; SameSite=Strict
X-Request-Id: [FILTERED]
X-Runtime: 0.781739
Content-Security-Policy-Report-Only: default-src https: 'self'; base-uri 'self'; child-src 'self'; connect-src 'self' wws:; font-src 'self' data:; form-action 'self' github.com; frame-ancestors 'self'; frame-src 'self'; img-src 'self' www.gravatar.com data:; media-src 'self'; object-src 'self'; plugin-types application/pdf; script-src 'self'; style-src 'self'; report-uri https://[FILTERED].report-uri.io/r/default/csp/enforce
Referrer-Policy: origin-when-cross-origin
X-Content-Type-Options: nosniff
X-Download-Options: noopen
X-Frame-Options: SAMEORIGIN
X-Permitted-Cross-Domain-Policies: none
X-XSS-Protection: 1; mode=block
Transfer-Encoding: chunked

**Comment by jacobbednarz** [comment]
Not sure if it is the cause but with the report-uri service, you have a 'reportOnly' and 'enforce' endpoint to recieve violations. In your issue above, you are using the 'enforce' endpoint for a report only CSP and HPKP directives.

**Comment by Xosmond** [comment]
Yes, that was the problem, maybe if you can explain when should i use enforce or reportOnly would be helpful. Anyway thank you.

**Comment by jacobbednarz** [comment]
report-uri.io is a separate service to this gem so it's probably best to get in touch with them regarding the correct (and expected) usage. My understanding is when your running in reporting mode, you use the 'reportOnly' endpoint however when you are enforcing the policy you swap to the 'enforce' endpoint.

**Comment by Xosmond** [comment]
Oh i see. Thank you, I'm new in this.

**Comment by oreoshake** [comment]
Ah, that's a relief! Thanks @jacobbednarz and I'm glad things are working now @Xosmond. 

I saw "rails 5" and something I couldn't explain and I started to get nervous 😰   

> My understanding is when your running in reporting mode, you use the 'reportOnly' endpoint however when you are enforcing the policy you swap to the 'enforce' endpoint.

I believe this is because there is no way to distinguish a violation report from `Content-Security-Policy` and `Content-Security-Policy-Report-Only` which creates the need for separate endpoints.
